### PR TITLE
Add comprehensive test suite for file persistence and change tracking

### PR DESCRIPTION
**Pull request by grouville**
# Summary

Comprehensive test suite to ensure reliable file persistence, change tracking, and environment isolation in container-use.

### Core Functionality Tests
- **Work Persistence**: Files and changes persist across container restarts
- **Change Tracking**: Automatic capture and auditing of all modifications
- **File Handling**: Graceful handling of Python caches, binaries, and large files
- **Environment Isolation**: Safe parallel operations with proper isolation
- **Configuration**: Reliable persistence of base images and setup commands

### Test Coverage
- Git operations (error handling, worktrees, empty directories)
- Selective staging for Python cache and binary files
- Configuration persistence and isolation verification
- Behavior-driven tests focused on user experience
- Test helpers for isolated environment setup

### Implementation Notes
- Modified `git.go` to support test isolation via `CONTAINER_USE_CONFIG_DIR`
- Documented known limitations with skipped tests
- Added to GitHub workflows for CI/CD

## Status
Ready for review

## UPDATE 06/25/2025

Rewrote the entire test suite, following the new paradigm. The idea is to have integration tests that mimick what a `user` will do to end up in a bug / weird situation

Already ready for review

Introduced the `Repository: basePath`: 
```shell
type Repository struct {
	userRepoPath string
	forkRepoPath string
	basePath     string // defaults to ~/.config/container-use if empty
}
```
instead of the `CONTAINER_USE_CONFIG_DIR` to isolate tests and run them in parallel 

The `WithRepository` is a helper function that spins up Dagger -- a repo with some configurable setup function and returns a userActions, which has access to the same function as the agent would. The idea is for this abstraction to also be able to mimick the CLI calls in the near future to declaratively express any user scenario

TODO: 
- [ ] do not copy the mcp handler
- [ ] cleanup a bit

### CODE REVIEW

File: environment/test_helpers.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
+package environment
```
Comment (cwlbraa) [inline environment/test_helpers.go]: does this break down if we put it in `package environment_test`?

File: environment/unit_test.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
+package environment
```
Comment (cwlbraa) [inline environment/unit_test.go]: the unit tests will conflict with #59, but not in a way that's hard to fix i don't think, just gotta move some shit around

File: environment/README.md [hunk 3]
```diff
@@ -71,3 +71,48 @@ So the flow is: **Branch** (the logical environment) → **Worktree** (filesyste
 - `environment.go` - Core environment management
 - `git.go` - Worktree and Git integration
 - `filesystem.go` - File operations within containers
+
+## Testing
+
+The environment package includes comprehensive unit and integration tests that verify critical user behaviors:
+
+### Running Tests
```
Comment (cwlbraa) [inline environment/README.md]: i think it's worth immediately wiring these up to CI/dagger prior to merging

File: environment/README.md [hunk 4]
```diff
@@ -71,3 +71,66 @@ So the flow is: **Branch** (the logical environment) → **Worktree** (filesyste
...
+- Automatic change tracking
+- Environment isolation
+- Edge case handling (Python cache, binary files, large files)
+- Performance with 100+ file projects
+- Configuration persistence
+
+### Known Issues
+
+The test suite documents several known bugs with skipped tests:
+- Python `__pycache__` directories cause git operations to fail
```
Comment (aluzzardi) [inline environment/README.md]: I'd either file these as issues or not list them at all, we'll forget to keep this README up to date :)

File: environment/README.md [hunk 5]
```diff
@@ -71,3 +71,66 @@ So the flow is: **Branch** (the logical environment) → **Worktree** (filesyste
 - `environment.go` - Core environment management
 - `git.go` - Worktree and Git integration
 - `filesystem.go` - File operations within containers
+
+## Testing
```
Comment (aluzzardi) [inline environment/README.md]: nit: Maybe move these to `CONTRIBUTING.md`?

File: environment/unit_test.go (hunk 2 above)
Comment (aluzzardi) [inline environment/unit_test.go]: nit: Go convention, unit tests should be `environment_test.go` if we're testing the environment file, not `unit_test.go`

File: environment/integration_test.go [hunk 6]
```diff
@@ -0,0 +1,729 @@
+package environment
```
Comment (aluzzardi) [inline environment/integration_test.go]: nit: MAYBE we should move these into their own package, e.g. `environment/integration` or something, keep the unit tests top-level. MAYBE. Thoughts?

**Review by aluzzardi — COMMENTED** [review]
Thank you @grouville, much appreciated! Some minor comments

File: environment/integration_test.go (hunk 6 above)
Comment (cwlbraa) [reply environment/integration_test.go]: imo for small projects the thing he's doing with -short is quite nice and provides a similar test grouping UX. i'd like to try it out and see if there are other drawbacks that aren't yet obvious to me... the one i can see from here is it might be too easy to share helpers that turn unit tests into int tests.

Comment (grouville) [reply environment/integration_test.go]: > nit: MAYBE we should move these into their own package, e.g. `environment/integration` or something, keep the unit tests top-level. MAYBE. Thoughts?

Regarding the integration tests: I’ve kept them in the environment package because they need access to the private functions to thoroughly verify the implementation. Moving them to a separate package (like environment_test or environment/integration) would limit testing to the public API only, reducing coverage (or requiring to turn more of those functions public. What would be your recommended approach (I don't mind) ?

File: environment/README.md (hunk 3 above)
Comment (grouville) [reply environment/README.md]: done ✅

File: environment/test_helpers.go (hunk 1 above)
Comment (grouville) [reply environment/test_helpers.go]: quite a lot -- as some functions i am using are not public. Happy to move them too https://github.com/dagger/container-use/pull/65#discussion_r2148877778

File: environment/README.md (hunk 4 above)
Comment (grouville) [reply environment/README.md]: ✅

File: environment/README.md (hunk 5 above)
Comment (grouville) [reply environment/README.md]: ✅

File: environment/unit_test.go (hunk 2 above)
Comment (grouville) [reply environment/unit_test.go]: ✅

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+   git clone git@github.com:dagger/container-use.git
+   ```
+3. **Install Dependencies**:
+
+   ```bash
+   go mod download
+   ```
+4. **Container Runtime**: Ensure you have a compatible container runtime installed (e.g., Docker).
+
+## Testing
```
Comment (cwlbraa) [inline CONTRIBUTING.md:38]: we should document building as well

```suggestion
## Building

```bash
go build ./cmd/cu`
```

## Testing
```

Comment (cwlbraa) [reply CONTRIBUTING.md:38]: lmao this is an impossible suggestion to make because the markdowns conflict, but you get the idea

File: .github/workflows/cu.yml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-name: Build
+name: Build and Test
```
Comment (cwlbraa) [inline .github/workflows/cu.yml]: can we have 2 workflow files so they show as separate rows in the checks panels? (both on commits and at the tail of the PR log)

File: CONTRIBUTING.md (hunk 7 above)
Comment (cwlbraa) [reply CONTRIBUTING.md:38]: this would also imply we can delete the build instructions from the main readme

Comment (grouville) [reply CONTRIBUTING.md:38]: ✅

File: .github/workflows/cu.yml (hunk 8 above)
Comment (grouville) [reply .github/workflows/cu.yml]: ✅

**Review by cwlbraa — APPROVED** [review]
lgtm, just need to decide whether this or #62 goes in first... conventional wisdom says tests go in first, but idk cc @aluzzardi

File: repository/repository.go [hunk 9]
```diff
@@ -25,12 +26,39 @@ const (
...
+func (r *Repository) getBasePath() string {
+	if r.basePath != "" {
+		return r.basePath
+	}
+	return cuGlobalConfigPath
+}
+
+// getRepoPath returns the path for storing repository data
+func (r *Repository) getRepoPath() string {
+	return r.getBasePath() + "/repos"
```
Comment (aluzzardi) [inline repository/repository.go:43]: nit: use `filepath.Join`

File: repository/repository.go [hunk 10]
```diff
@@ -25,12 +26,39 @@ const (
...
+}
+
+// getWorktreePath returns the path for storing worktrees
+func (r *Repository) getWorktreePath() string {
+	return r.getBasePath() + "/worktrees"
```
Comment (aluzzardi) [inline repository/repository.go:48]: `filepath.Join`

File: repository/repository.go [hunk 11]
```diff
@@ -25,12 +26,39 @@ const (
...
+
+// OpenWithBasePath opens a repository with a custom base path for container-use data.
+// This is useful for tests that need isolated environments.
+func OpenWithBasePath(ctx context.Context, repo string, basePath string) (*Repository, error) {
+	output, err := RunGitCommand(ctx, repo, "rev-parse", "--show-toplevel")
 	if err != nil {
-		if strings.Contains(err.Error(), "not a git repository") {
+		// Check for exit code 128 which means not a git repository
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) && exitErr.ExitCode() == 128 {
```
Comment (aluzzardi) [inline repository/repository.go:62]: nice!

Comment (aluzzardi) [reply repository/repository.go:62]: For **later** as a follow-up: **maybe** we should add this logic to `RunGitCommand` directly, catch all git commands that are not run from a git repository?

Although I think it's impossible since everything starts with `Open()`

File: repository/git.go [hunk 12]
```diff
@@ -482,13 +485,16 @@ func (r *Repository) isBinaryFile(worktreePath, fileName string) bool {
...
+	origin, err := RunGitCommand(ctx, repo, "remote", "get-url", "origin")
 	if err != nil {
 		// If not -- this repository is a local one, we're going to use the filesystem path for the container-use repo
-		if strings.Contains(err.Error(), "No such remote") {
-			return homedir.Expand(filepath.Join(cuRepoPath, repo))
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) && exitErr.ExitCode() == 2 {
+			// Exit code 2 means the remote doesn't exist
+			// Use just the base name for temporary directories to avoid deeply nested paths
+			return homedir.Expand(filepath.Join(r.getRepoPath(), filepath.Base(repo)))
```
Comment (aluzzardi) [inline repository/git.go]: So actually that was done on purpose -- we used to use the directory name, but that failed if multiple repositories had the same name on different directories (e.g. `~/work/foo` and `~/Documents/foo`).

Was there an issue this is fixing, or was it for aesthetics?

File: repository/repository.go [hunk 13]
```diff
@@ -25,12 +26,39 @@ const (
 type Repository struct {
 	userRepoPath string
 	forkRepoPath string
+	basePath     string // defaults to ~/.config/container-use if empty
+}
+
+// getBasePath returns the base path for container-use data, using the default if not set
+func (r *Repository) getBasePath() string {
```
Comment (aluzzardi) [inline repository/repository.go:34]: nit: This logic is already handled by `Open()` (e.g. setting `cuGlobalConfigPath` when calling `Open()`).

I think it would be simpler to avoid duplication of this logic, always assume `r.basePath` is set, and get rid of `getBasePath` altogether

File: environment/integration/repository_test.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
+package integration
```
Comment (aluzzardi) [inline environment/integration/repository_test.go:1]: Shouldn't these go into `repository/repository_test.go`? We're testing `Open` over there and the rest here here

**Comment by cwlbraa** [comment]
> To me integration tests would be like ... testing cu with claude for instance.

i'd prolly call tests involving claude e2e, but if we're integrating anything here it's probably the MCP server against repository+env
